c2rust-analyze/src/c_void_casts.rs

```diff
@@ -0,0 +1,484 @@
+        use CVoidCastDirection::*;
+        match direction {
+            From => &self.from,
+            To => &self.to,
+        }
+    }
+
+    fn direction_mut(
+        &mut self,
+        direction: CVoidCastDirection,
+    ) -> &mut CVoidCastsUniDirectional<'tcx> {
+        use CVoidCastDirection::*;
+        match direction {
+            From => &mut self.from,
+            To => &mut self.to,
+        }
+    }
+
+    /// See [`CVoidCastsUniDirectional::get_adjusted_place_or_default_to`].
+    pub fn get_adjusted_place_or_default_to(
+        &self,
+        loc: Location,
+        direction: CVoidCastDirection,
+        place: Place<'tcx>,
+    ) -> Place<'tcx> {
+        self.direction(direction)
+            .get_adjusted_place_or_default_to(loc, place)
+    }
+
+    /// See [`CVoidCastsUniDirectional::insert_cast`].
+    fn insert_cast(&mut self, direction: CVoidCastDirection, loc: Location, cast: CVoidCast<'tcx>) {
+        self.direction_mut(direction).insert_cast(loc, cast)
+    }
+
+    /// See [`CVoidCastsUniDirectional::insert_call`].
+    fn insert_call(&mut self, direction: CVoidCastDirection, loc: Location, cast: CVoidCast<'tcx>) {
+        self.direction_mut(direction).insert_call(loc, cast)
+    }
+
+    /// Determine if this [`Statement`] should be skipped
+    /// because it contains a [`CVoidCast`] that is in this [`CVoidCasts`],
+    /// as it will be handled in a way that effectively elides the cast.
+    ///
+    /// That is, if this [`Statement`] is an [`Assign`],
+    /// then it will be skipped if:
+    ///
+    /// * the lhs [`Place`] of the [`Assign`] is a [`CVoidPtr`] in the [`To`] direction,
+    ///     as in the [`To`] direction, the [`CVoidPtr`] is on the lhs.
+    ///
+    /// * the rhs [`Place`] of the [`Cast`] of the [`Assign`] is a [`CVoidPtr`] in the [`From`] direction,
+    ///     as in the [`From`] direction, the [`CVoidPtr`] is on the rhs.
+    ///
+    /// [`Assign`]: rustc_middle::mir::StatementKind::Assign
+    /// [`Cast`]: rustc_middle::mir::Rvalue::Cast
+    /// [`From`]: CVoidCastDirection::From
+    /// [`To`]: CVoidCastDirection::To
+    pub fn should_skip_stmt(&self, loc: Location) -> bool {
+        self.to.casts.contains_key(&loc) || self.from.casts.contains_key(&loc)
+    }
+
+    /// Checking whether a statement could modify a particular Place is
+    /// nontrivial (particularly if the Place involves Deref projections),
+    /// but currently rustc always uses a temporary for the LHS of the
+    /// cast in the desired free(ptr as *mut c_void) pattern, so I think we
+    /// can conservatively treat the Place as unmodified only if (1) it's
+    /// just a Local, with no projections, and (2) that local isn't
+    /// mentioned in the LHS or RHS of any statements between the cast and
+    /// the call.
+    fn is_place_modified_by_statement(p: &Place, stmt: &Statement) -> bool {
+        if !p.projection.is_empty() {
+            return false;
```

Comment:
```suggestion
            return true;
```

A `Place` is unmodified if (1) and (2), where (1) is a `Local` with no projections.  So if a `Place` has projections (`!p.projection.is_empty()`), the `Place` is considered modified, so this should `return true` here.